### DISCUSSION TITLE
router.query objects remains empty when site is hosted on ubuntu server using nginx.

### DISCUSSION THREAD

**Discussion by siddhesh-nalawade** [vercel/next.js#40830] [Help]
I have hosted next js app on ubuntu server using nginx. I have used dynamic routing in my application. When I pass the params using url I saw that router.query object remains empty. That is why useState is not updating the State of app. But when I click on the refresh page I get the params in query.router object and also State of app gets updated. The app is not facing this issue on local host. 

See below screenshot I am getting empty ({}) router.query object. I checked whether I am getting something or not in url using window.location.url. I can see the app id which is passed through url i.e 168.
![t](https://user-images.githubusercontent.com/97395597/191911765-65c92c3e-3729-43ca-96fc-fc2f80a79739.png)

**Comment by icyJoseph** [comment]
As per docs:

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
> 
> After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

Though, initially the query params might be empty, after the initial render, the query should become available. From your description, there must be something else going on. 

Could you try to log `useRouter().query` on a page that doesn't use dynamic routing, for example your landing page? `https://my-domain.com/?foo=bar`, what does it log?

Does this happen also if you `next build && next start` locally? Are you able to setup your NGINX server locally as well? It'd be nice to debug locally, on the same conditions as the production environment.

**Reply by siddhesh-nalawade** [reply]
> Could you try to log useRouter().query on a page that doesn't use dynamic routing, for example your landing page? https://my-domain.com/?foo=bar, what does it log?

It is printing correct query params without page refresh.
![image](https://user-images.githubusercontent.com/97395597/191917382-e86813cc-77ff-4eb3-ba31-41000971fb22.png)

> Does this happen also if you next build && next start locally? Are you able to setup your NGINX server locally as well? It'd be nice to debug locally, on the same conditions as the production environment.

I have not tried it locally.

**Reply by icyJoseph** [reply]
Ok, but the query updating mechanism does work. 

Are you perhaps doing something like:

```tsx
const MyComponent = () => {
 const {query} = useRouter();
 const [state, setState] = useState(query);
 return <div>hi</div>
}
```

~~This will seemingly work during development mode I think, because of `StrictMode`, https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state, though I need to verify that.~~ No, I was wrong, because of how the query updating mechanism works with Strict Mode, in the above, `state` will always be `{}`, in dev mode.

In production, because the `useState` initialiser is run only once, when the query updates, it won't trigger an update. I am guessing here though, I am not sure this is your actual issue. Would be good if you can test with `next build && next start` without the server, perhaps it is enough to trigger the error? - This still applies, but I don't think it is your issue. 

In either case, `query` does update as expected, first empty, and then the values appear.

Got a repository to look at?

**Reply by icyJoseph** [reply]
Also, one thing I am trying to get through is that, these things update client side, so why would the NGINX server have anything to do with it? But, you know, in software development, I've been surprised more than once before.

**Reply by siddhesh-nalawade** [reply]
> Does this happen also if you next build && next start locally? Are you able to setup your NGINX server locally as well? It'd be nice to debug locally, on the same conditions as the production environment

I have created similar setup on my localhost. I am now facing similar issue on localhost as well.

> Got a repository to look at?

Sorry the repo is private. Can't give access to anyone.  But I can give you the code snippit of this page.
```

import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
import Auth from "../../../components/Auth/Auth";
import C4DAppSections from "../../../components/C4DAppSections/C4DAppSections";
import BehaviourData from "../../../components/ShowData/BehaviourData";
import CarerStoryData from "../../../components/ShowData/CarerStoryData";
import CarerSupportData from "../../../components/ShowData/CarerSupportData";
import GCSPData from "../../../components/ShowData/GCSPData";

function AppDetails() {
  const router = useRouter();
  const [appId, setAppId] = useState("");
  const [appName, setAppName] = useState("");
  const [currentC4dSecton, setcurrentC4dSecton] = useState("Carer Story");
  const [c4dApp, setc4dApp] = useState([
    "Carer Story",
    "Carer Support",
    "General Consideration and Safety Precautions",
    "Behaviour Section",
  ]);

  const getCurrentC4dSecton = () => {
    let currentC4dComponentName;
    switch (currentC4dSecton) {
      case "Carer Story":
        currentC4dComponentName = <CarerStoryData appId={appId} />;
        break;
      case "Carer Support":
        currentC4dComponentName = <CarerSupportData appId={appId} />;
        break;
      case "General Consideration and Safety Precautions":
        currentC4dComponentName = <GCSPData appId={appId} />;
        break;
      case "Behaviour Section":
        currentC4dComponentName = <BehaviourData appId={appId} />;
        break;

      default:
        currentC4dComponentName = <CarerStoryData appId={appId} />;
        break;
    }
    return (
      <div
        className=" justify-content-center shadow p-4 mt-3"
        style={{ border: "2px solid blue", borderRadius: "20px" }}
      >
        <h1>{currentC4dSecton}</h1>
        {currentC4dComponentName}
      </div>
    );
  };

  const changeCurrentSection = (type) => {
    setcurrentC4dSecton(type);
  };

  useEffect(() => {
    setAppId(router.query.id);
    setAppName(router.query.app);
  }, [router]);

  return (
    <>
      <h1>{appName}</h1>
      <div className="row">
        {c4dApp.map((c4d_name) => (
          <div className="col-3" key={c4d_name}>
            <C4DAppSections
              isSelected={c4d_name === currentC4dSecton}
              c4d_name={c4d_name}
              app_id={appId}
              changeCurrentSection={changeCurrentSection}
            />
          </div>
        ))}
      </div>
      <hr />
      <Auth>{getCurrentC4dSecton()}</Auth>
    </>
  );
}

export default AppDetails;
```

**Reply by icyJoseph** [reply]
Ok, but we now know the problem is not the NGINX setup but something that's triggering at build time, correct? I'll take a look at the snippet, see if there's anything that pops-up.

### EDIT

Btw, which Next.js version are you using? Do you have middleware? Perhaps you are doing something in `_app`?

**Reply by siddhesh-nalawade** [reply]
I am using Next.js v12.3.0. I am not using middleware.

_app.js
```

import "bootstrap/dist/css/bootstrap.css";
import "../styles/styles.css";
import Script from "next/script";
import Layout from "../components/LayoutComponent/LayoutComponent";
import { SessionProvider } from "next-auth/react";

function MyApp({ Component, pageProps: { session, ...pageProps } }) {
  return (
    <>
      <SessionProvider session={pageProps.session} basePath="/my_app_name/api/auth">
        <Layout>
          <Component {...pageProps} />
        </Layout>
      </SessionProvider>

      <Script
        src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/js/bootstrap.bundle.min.js"
        integrity="sha384-ygbV9kiqUc6oa4msXn9868pTtWMgiQaeYH7/t7LECLbyPA2x65Kgf80OJFdroafW"
        crossOrigin="anonymous"
      />
    </>
  );
}

export default MyApp;

```

**Reply by icyJoseph** [reply]
Well, I don't see anything that'd me raise an eyebrow. 

- Have we discarded the NGINX as source of the problem?
- Are you able to make a [https://stackblitz.com/](https://stackblitz.com/) reproduction?

One thing I may have misunderstood. Loading the page directly on `AppDetails` works, but the problem is when you navigate to the page, from another page right?

**Reply by siddhesh-nalawade** [reply]
> Have we discarded the NGINX as source of the problem?

I think NGINX might be cause of issue because every thing works fine on local host but when I host site on server or on local machine issue arises.

> One thing I may have misunderstood. Loading the page directly on AppDetails works, but the problem is when you navigate to the page, from another page right?

Yes. When I naviagate from one page to another data not loads at first time but If I refresh the page then data loads.

**Reply by icyJoseph** [reply]
Alright, wow this is tougher than I thought it'd be!

> Yes. When I navigate from one page to another data not loads at first time but If I refresh the page then data loads.

This is what makes this so strange for me, because this transaction shouldn't include the server at all. And the fact that you say it breaks when NGINX is involved, but not if you just start it and run normal, makes it even worse.

Perhaps we need to use Devtools here, if you could try to find the Router Context Provider on the built version of your APP, with NGiNX, locally, does it contain the query parameters?

<img width="770" alt="Screenshot 2022-09-26 at 10 06 43" src="https://user-images.githubusercontent.com/21013447/192225443-a4db0395-8ed3-4eb9-aecb-585e292e62b1.png">
